src/protections/DetectFederationLag.ts

```diff
@@ -0,0 +1,744 @@
+            this.cleanup(now);
+        }
+        if (this.settings.ignoreRooms.value.has(roomId)) {
+            // Room is ignored.
+            return;
+        }
+        const sender = event['sender'] as string;
+        if (typeof sender !== "string") {
+            // Ill-formed event.
+            return;
+        }
+        if (sender === await mjolnir.client.getUserId()) {
+            // Let's not create loops.
+            return;
+        }
+        const domain = new UserID(sender).domain;
+        if (!domain) {
+            // Ill-formed event.
+            return;
+        }
+
+        const origin = event['origin_server_ts'] as number;
+        if (typeof origin !== "number") {
+            // Ill-formed event.
+            return;
+        }
+        const delay = now.getTime() - origin;
+        if (delay < 0) {
+            // Could be an ill-formed event.
+            // Could be non-motonic clocks or other time shennanigans.
+            return;
+        }
+
+        let roomInfo = this.lagPerRoom.get(roomId);
+        if (!roomInfo) {
+            roomInfo = new RoomInfo(now);
+            this.lagPerRoom.set(roomId, roomInfo);
+        }
+
+        const localDomain = new UserID(await mjolnir.client.getUserId()).domain
+        const isLocalDomain = domain === localDomain;
+        const thresholds =
+            isLocalDomain
+                ? {
+                    enterWarningZone: this.settings.localHomeserverLagEnterWarningZoneMS.value,
+                    exitWarningZone: this.settings.localHomeserverLagExitWarningZoneMS.value,
+                }
+                : {
+                    enterWarningZone: this.settings.federatedHomeserverLagEnterWarningZoneMS.value,
+                    exitWarningZone: this.settings.federatedHomeserverLagExitWarningZoneMS.value,
+                };
+
+        const diff = roomInfo.pushLag(domain, delay, this.latestHistogramSettings, thresholds, now);
+        if (diff === AlertDiff.NoChange) {
+            return;
+        }
+
+        if (roomInfo.latestWarning.getTime() + this.settings.warnAgainAfterMS.value > now.getTime()) {
+            if (!isLocalDomain || diff !== AlertDiff.Start) {
+                // No need to check for alarms, we have raised an alarm recently.
+                return;
+            }
+        }
+
+        // Check whether an alarm needs to be raised!
+        const isLocalDomainOnAlert = roomInfo.isServerOnAlert(localDomain);
+        if (roomInfo.alerts > this.settings.numberOfLaggingFederatedHomeserversEnterWarningZone.value
+            || isLocalDomainOnAlert) {
+            // Raise the alarm!
+            if (!roomInfo.latestAlertStart) {
+                roomInfo.latestAlertStart = now;
+            }
+            roomInfo.latestAlertStart = now;
+            // Background-send message.
+            const stats = roomInfo.globalStats();
+            /* do not await */ logMessage(LogLevel.WARN, "FederationLag",
```

Comment:
There is always a risk, but if it hangs/timeouts here, it will also probably hang/timeout at the next line.

The reason for which I didn't `await` is simply because we don't really need to be certain that the message has been properly sent before proceeding.